Framework/HistogramData/src/Interpolate.cpp

```diff
@@ -0,0 +1,209 @@
+ * @param minCalculated The minimum number of calculated values required
+ * by the routine
+ */
+void sanityCheck(const Histogram &input, const size_t stepSize,
+                 const size_t minCalculated, const char *method) {
+  if (input.yMode() == Histogram::YMode::Uninitialized) {
+    throw std::runtime_error(
+        "interpolate - YMode must be defined for input histogram.");
+  }
+  const auto ysize = input.y().size();
+  if (stepSize >= ysize) {
+    throw std::runtime_error("interpolate - Step size must be smaller "
+                             "than the number of points");
+  }
+  // First and last points are always assumed to be calculated
+  const size_t ncalc = numberCalculated(ysize, stepSize);
+  if (ncalc < minCalculated) {
+    std::ostringstream os;
+    os << "interpolate - " << method << " requires " << minCalculated
+       << " points but only " << ncalc << " were found.";
+    throw std::runtime_error(os.str());
+  }
+}
+
+/**
+ * Perform linear interpolation assuming the input data set to be a Histogram
+ * with XMode=Points. It is assumed all sanity cheks have been performed
+ * by the interpolateLinear entry point.
+ * @param input See interpolateLinear
+ * @param stepSize See interpolateLinear
+ * @param ynew A reference to the output Y values
+ * @return See interpolateLinear
+ */
+void interpolateYLinearInplace(const Histogram &input, const size_t stepSize,
+                               HistogramY &ynew) {
+  auto xold = input.points();
+  auto &yold = input.y();
+  auto nypts = yold.size();
+  size_t step(stepSize), index2(0);
+  double x1(0.), x2(0.), y1(0.), y2(0.), overgap(0.);
+  // Copy over end value skipped by loop
+  ynew.back() = yold.back();
+  for (size_t i = 0; i < nypts - 1; ++i) // Last point has been calculated
+  {
+    double xp = xold[i];
+    if (step == stepSize) {
+      x1 = xp;
+      index2 = ((i + stepSize) >= nypts ? nypts - 1 : (i + stepSize));
+      x2 = xold[index2];
+      overgap = 1.0 / (x2 - x1);
+      y1 = yold[i];
+      y2 = yold[index2];
+      step = 1;
+      ynew[i] = yold[i];
+      continue;
+    }
+    // Linear interpolation
+    ynew[i] = (xp - x1) * y2 + (x2 - xp) * y1;
+    ynew[i] *= overgap;
+    step++;
+  }
+}
+/**
+ * Perform cubic spline interpolation. It is assumed all sanity checks have been
+ * performed by the interpolateCSpline entry point.
+ * @param input See interpolateCSpline
+ * @param stepSize See interpolateCSpline
+ * @param ynew A reference to the output Y values
+ * @return See interpolateCSpline
+ */
+void interpolateYCSplineInplace(const Histogram &input, const size_t stepSize,
+                                HistogramY &ynew) {
+  auto xold = input.points();
+  auto &yold = input.y();
+  auto nypts = yold.size();
+
+  const auto ncalc = numberCalculated(nypts, stepSize);
+  std::vector<double> xc(ncalc), yc(ncalc);
+  for (size_t step = 0, i = 0; step < nypts; step += stepSize, i += 1) {
+    xc[i] = xold[step];
```

Comment:
In this case I want `xc` to just contain the pre-calculated points and so it's smaller than `xold`.I don't know if it might be confusing to try and orchestrate that with the cow pointers. I'll have a look.